runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java

```diff
@@ -79,24 +79,44 @@
         }
     }
 
+    public static boolean XWalkLibraryCompressed(Context context) {
+        if (context == null) return false;
+
+        return XWalkCompressUtil.XWalkLibraryCompressed(context, MANDATORY_LIBRARIES);
+    }
+
+    public static void decompressXWalkLibrary(Context context) throws Exception {
+        if (context == null) return;
+
+        String lib = PathUtils.getDataDirectory(context.getApplicationContext());
+        long start = System.currentTimeMillis();
+        XWalkCompressUtil.decompressXWalkLibrary(context, MANDATORY_LIBRARIES, lib);
+        long end = System.currentTimeMillis();
+        Log.d(TAG, "decompress library cost: " + (end - start) + " milliseconds.");
+    }
+
     public static void loadXWalkLibrary(Context context) throws UnsatisfiedLinkError {
-        if (sLibraryLoaded) return;
-
-        // If context is null, it's called from wrapper's ReflectionHelper to try
-        // loading native library within the package. No need to try load from library
-        // package in this case.
-        // If context's applicationContext is not the same package with itself,
-        // It's a cross package invoking, load core library from library apk.
-        // Only load the native library from /data/data if the Android version is
-        // lower than 4.2. Android enables a system path /data/app-lib to store native
-        // libraries starting from 4.2 and load them automatically.
-        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR1 && context != null &&
-                !context.getApplicationContext().getPackageName().equals(context.getPackageName())) {
-            for (String library : MANDATORY_LIBRARIES) {
-                System.load("/data/data/" + context.getPackageName() + "/lib/" + library);
+        if (sLibraryLoaded || (context == null)) return;
+
+        if (XWalkLibraryCompressed(context)) {
+            String lib = PathUtils.getDataDirectory(context.getApplicationContext());
+            if (lib != null) {
+                for (String library : MANDATORY_LIBRARIES) {
+                    System.load(lib + "/" + library);
+                }
+            }
```

Comment:
If the so file is compressed, is it necessary to call System.load for each file? Why not just throw an exception?